src/framework/constraints.ml

```diff
@@ -497,6 +497,90 @@ struct
   let event (ctx:(D.t,G.t,C.t,V.t) ctx) (e:Events.t) (octx:(D.t,G.t,C.t,V.t) ctx):D.t = lift_fun ctx D.lift S.event ((|>) (conv octx) % (|>) e) `Bot
 end
 
+module NoContext = struct let name = "no context" end
+module IntConf = 
+struct
+  let n () = get_int "ana.context.ctx_gas_value" + 1
+  let names x = Format.asprintf "%d" x
+end
+
+(** Lifts a [Spec] with the context gas variable. For every function call the context gas is reduced. 
+    If the context gas is 0, the remaining function calls are analyzed context insensitively (before the analysis is context sensitive)*)
+module ContextGasLifter (S:Spec)
```

Comment:
The way I understand the current implementation, the context gas is global here in the sense that it limits the general depth of the callstack? is that understanding correct?
Have you considered having a per-function context gas? This would tame recursion without losing precision for non-recursive programs...

Reply:
Yes, that should be correct. The context gas is always decreased when entering a function, but inside a function, the value stays the same. 
For example, if `main `has context gas `5` and calls function `f`, and function `g`, both functions are called with a context gas value of `4`. 

I also thought of that. Would this per-function context gas be globally, so across different paths? I could imagine that the accuracy is less balanced across paths... But I will discuss this approach with Julian in our meeting today :)

Reply:
> If the context gas is 0, the remaining function calls are analyzed context insensitively (before the analysis is context sensitive)

This also seems surprising: when context gas is enabled, all other context-sensitivity configuration is silently ignored.
I would expect it to degenerate to whatever contexts `S.C` specifies. One can get this behavior by just making everything in `S.C` context-insensitive, but would be more general.

This also begs the question, why is this a `Spec` lifter, while the callstrings are actual analyses? It seems to me that this is also a mere stack abstraction, where the stack is abstracted just to its size, bounded by gas (which would collapse all deeper calls together when gas runs out). This is basically `Lattice.Chain`.

Reply:
The idea behind the context gas approach is to use the normal context-sensitive analyzer of Goblint for all function calls that have a gas value greater zero. In case the gas value is zero, the function does not receive any context information (we realize this by passing `NULL` as context). It may happen that multiple function calls to the same function receive no context and are merged for the analysis. Consequently, the context-sensitivity configuration is not ignored if the gas is greater zero. And since we need to switch from passing the sensitive contexts or passing `NONE` depending on the gas value, it is necessary to implement is as lifter since we need to overwrite the context information of the analyses. 
In case the context gas value never reaches zero, the context gas analysis is equal to the standard configuration of Goblint. 

For example, we have the call-sequence `main -> f(2) -> f(1) -> f(0)` and a gas of `3`. In this case, `main` would be analyzed with gas `2´ and therefore context-sensitive. `f(2)` receives gas `1`, and is also analyzed as usual. For `f(1)` the gas reaches zero, and consequently, this call is analyzed without context information. The same happens for `f(0)`, and consequently, they have the same context and are analyzed together. 

I hope this answers all parts of the question??

Reply:
c.f. https://github.com/tum-cit-pl/taming-contexts for a formal description